universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -258,25 +259,40 @@ public ApplicantProgramBlocksController(
 
               Optional<String> bucket = request.queryString("bucket");
               Optional<String> key = request.queryString("key");
+
+              // Original file name is only set for Azure, where we have to generate a UUID when
+              // uploading a file to Azure Blob storage because we cannot upload a file without a
+              // name. For AWS, the file key and original file name are the same
+              Optional<String> originalFileName = request.queryString("originalFileName");
+
               if (!bucket.isPresent() || !key.isPresent()) {
                 return failedFuture(
                     new IllegalArgumentException("missing file key and bucket names"));
               }
 
-              String applicantFileUploadQuestionKeyPath =
+              FileUploadQuestion fileUploadQuestion =
                   block.get().getQuestions().stream()
                       .filter(question -> question.getType().equals(QuestionType.FILEUPLOAD))
                       .findAny()
                       .get()
-                      .createFileUploadQuestion()
-                      .getFileKeyPath()
-                      .toString();
-              ImmutableMap<String, String> formData =
-                  ImmutableMap.of(applicantFileUploadQuestionKeyPath, key.get());
+                      .createFileUploadQuestion();
+
+              String applicantFileUploadQuestionKeyPath =
+                  fileUploadQuestion.getFileKeyPath().toString();
+              String applicantFileUploadQuestionOriginalFileNamePath =
+                  fileUploadQuestion.getOriginalFileNamePath().toString();
+
+              ImmutableMap.Builder<String, String> builder =
+                  new ImmutableMap.Builder<String, String>();
+              builder.put(applicantFileUploadQuestionKeyPath, key.get());
+              if (originalFileName.isPresent()) {
+                builder.put(
+                    applicantFileUploadQuestionOriginalFileNamePath, originalFileName.get());
+              }
```

Comment:
praise: great work figuring out how to handle optionally adding originalFileNames to the formData!